### ISSUE TITLE
False positives (FP) and true positives (TP) in object detection in tensorflow

### ISSUE THREAD

**Issue by jhonjam** [tensorflow/models#8412]
hello, i implemented the ssd detector in the tensorflow api, i get some metrics as mAP and loss. but I would like to obtain the number of true positives (TP), false positives (FP) and true negatives (TN) for each class, and from there, to be able to graph the PR curve.

I would appreciate if someone could guide me
thanks

**Comment by saberkun** [comment]
Please use the right template. object detection is the research model right?

**Comment by jhonjam** [comment]
> Please use the right template. object detection is the research model right?

Yes. I am tô used model object detection from api tensor flow.

**Comment by yang07ly** [comment]
+1. I'd greatly appreciate this feature because I need this for my research project! Or can anyone offer any insights on obtaining these metrics?

**Comment by jhonjam** [comment]
Hello, I solved the problem by modifying the code available [in](https://towardsdatascience.com/confusion-matrix-in-object-detection-with-tensorflow-b9640a927285)

In which I generate two .txt files; one referring to the predictions made by the model and the other to the ground truth (name2="TXTPREDICIONRCNN1/Im{}.txt".format(conttpredic), name="TXTVERDAFUNDAMENTALRCNN/Im{}.txt".format(contt))

This modification will generate N .txt files, which will contain 5 values, referring to the class, confidence level, and bounding box coordinates.

finally use the code available [in](https://github.com/Cartucho/mAP), which generates the FN, FP, mAP, etc.



`def process_detections(detections_record, categories):
    #dicdetec={}
    #dicdetec= dicDetec(detections_record, categories)

    record_iterator = tf.python_io.tf_record_iterator(path=detections_record)
    data_parser = tf_example_parser.TfExampleDetectionAndGTParser()

    confusion_matrix = np.zeros(shape=(len(categories) + 1, len(categories) + 1))
    #confusion_matrix = np.zeros(shape=(len(categories) , len(categories) ))

    image_index = 0
    contt=0
    conttpredic = 0
    image_y={}
    dicci={}
    boxes=[]
    boxescom=[]
    boxesdir={}
    boxesSc=[]
    boxescomSc=[]

    for string_record in record_iterator:
        example = tf.train.Example()
        example.ParseFromString(string_record)
        decoded_dict = data_parser.parse(example)
        
        image_index += 1
        
        if decoded_dict:
            groundtruth_boxes = decoded_dict[standard_fields.InputDataFields.groundtruth_boxes]
            groundtruth_classes = decoded_dict[standard_fields.InputDataFields.groundtruth_classes]
            
            name="TXTVERDAFUNDAMENTALRCNN/Im{}.txt".format(contt)

            vc=0
            f,c=groundtruth_boxes.shape
            file = open(name, "w", encoding="utf-8")

            for i in range(f):
                file.write(str(groundtruth_classes[i])+' ')
                for j in range(c):
                    file.write(str(groundtruth_boxes[i,j])+' ')
                    boxes.append(groundtruth_boxes[i,j])
                    vc+=1
                    if vc==4:
                        boxescom.append(boxes)
                        boxes=[]
                        file.write('\n')
                        vc=0
           
            print('lenboxes  '+str(len(boxescom)))
            boxesdir['boxes']=boxescom
         
            file.close()

            contt+=1
            

            detection_scores = decoded_dict[standard_fields.DetectionResultFields.detection_scores]
            detection_classes = decoded_dict[standard_fields.DetectionResultFields.detection_classes][detection_scores >= CONFIDENCE_THRESHOLD]
            detection_boxes = decoded_dict[standard_fields.DetectionResultFields.detection_boxes][detection_scores >= CONFIDENCE_THRESHOLD]
            
            
            name2="TXTPREDICIONRCNN1/Im{}.txt".format(conttpredic)
            vc2=0
            f2,c2=detection_boxes.shape
            file = open(name2, "w", encoding="utf-8")

            #print(''detection_boxes)

            for i in range(f2):
                file.write(str(detection_classes[i])+' ')
                for j in range(c2):
                    #print(groundtruth_boxes[i,j])
                    file.write(str(detection_boxes[i,j])+' ')
                    vc2+=1
                    if vc2==4:
                        file.write('\n')
                        vc2=0
            file.close()
            conttpredic+=1

           
            
            matches = []
            falso={}
                      
            falso = get_single_image_results(groundtruth_boxes, detection_boxes, 0.5)


            if image_index % 100 == 0:
                print("Processed %d images" %(image_index))
            
            for i in range(len(groundtruth_boxes)):
                for j in range(len(detection_boxes)):
                    iou = calc_iou(groundtruth_boxes[i], detection_boxes[j])
                    #print('groundtruth_boxes[i] '+ str(groundtruth_boxes[i]))
                   # print('detection_boxes[j] '+ str(detection_boxes[j]))
                    

                    if iou > IOU_THRESHOLD:
                        matches.append([i, j, iou])
                        #print(matches)
                    
            matches = np.array(matches)
            
            if matches.shape[0] > 0:
                # Sort list of matches by descending IOU so we can remove duplicate detections
                # while keeping the highest IOU entry.
                matches = matches[matches[:, 2].argsort()[::-1][:len(matches)]]
                # Remove duplicate detections from the list.
                matches = matches[np.unique(matches[:,1], return_index=True)[1]]
                # Sort the list again by descending IOU. Removing duplicates doesn't preserve
                # our previous sort.
                matches = matches[matches[:, 2].argsort()[::-1][:len(matches)]]
                # Remove duplicate ground truths from the list.
                matches = matches[np.unique(matches[:,0], return_index=True)[1]]
                #print('cinco '+str(matches))

            for i in range(len(groundtruth_boxes)):
                if matches.shape[0] > 0 and matches[matches[:,0] == i].shape[0] == 1:
                    confusion_matrix[groundtruth_classes[i] - 1][detection_classes[int(matches[matches[:,0] == i, 1][0])] - 1] += 1 
                else:
                    confusion_matrix[groundtruth_classes[i] - 1][confusion_matrix.shape[1] - 1] += 1
                    
            for i in range(len(detection_boxes)):
                if matches.shape[0] > 0 and matches[matches[:,1] == i].shape[0] == 0:
                    confusion_matrix[confusion_matrix.shape[0] - 1][detection_classes[i] - 1] += 1
            print(confusion_matrix)
        else:
            print("Skipped image %d" % (image_index))

    print("Processed %d images" % (image_index))

    

    return confusion_matrix`
